internal/appsec/limiter_test.go

```diff
@@ -0,0 +1,251 @@
+			ticksChan <- startTime
+			require.Equalf(t, int64(100), atomic.LoadInt64(&l.tokens), "Bucket should have exactly 100 tokens")
+		}
+	})
+
+	t.Run("bucket-refill-long", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+
+		for i := 0; i < 1000; i++ {
+			startTime = startTime.Add(3 * time.Second)
+			ticksChan <- startTime
+		}
+		require.Equalf(t, int64(100), atomic.LoadInt64(&l.tokens), "Bucket should have exactly 100 tokens")
+	})
+}
+
+func TestLimiter(t *testing.T) {
+	t.Run("allow-after-stop", func(t *testing.T) {
+		l := NewTokenTicker(3, 3)
+		l.Start()
+		require.True(t, l.Allow())
+		l.Stop()
+		// The limiter keeps allowing until there's no more tokens
+		require.True(t, l.Allow())
+		require.True(t, l.Allow())
+		require.False(t, l.Allow())
+	})
+
+	t.Run("concurrency", func(t *testing.T) {
+		//Tests the limiter's ability to sample the traces when subjected to a continuous flow of requests
+		//Each goroutine will continuously call the WAF and the rate limiter for 1 second
+		for nbUsers := 1; nbUsers <= 1000; nbUsers *= 10 {
+			t.Run(fmt.Sprintf("continuous-requests-%d-users", nbUsers), func(t *testing.T) {
+				var startBarrier, stopBarrier sync.WaitGroup
+				// Create a start barrier to synchronize every goroutine's launch and
+				// increase the chances of parallel accesses
+				startBarrier.Add(1)
+				// Create a stopBarrier to signal when all user goroutines are done.
+				stopBarrier.Add(nbUsers)
+				skipped := int64(0)
+				kept := int64(0)
+				l := NewTokenTicker(0, 100)
+
+				for n := 0; n < nbUsers; n++ {
+					go func(l Limiter, kept *int64, skipped *int64) {
+						startBarrier.Wait()      // Sync the starts of the goroutines
+						defer stopBarrier.Done() // Signal we are done when returning
+
+						for tStart := time.Now(); time.Since(tStart) < 1*time.Second; {
+							if !l.Allow() {
+								atomic.AddInt64(skipped, 1)
+							} else {
+								atomic.AddInt64(kept, 1)
+							}
+						}
+					}(l, &kept, &skipped)
+				}
+
+				l.Start()
+				defer l.Stop()
+				start := time.Now()
+				startBarrier.Done() // Unblock the user goroutines
+				stopBarrier.Wait()  // Wait for the user goroutines to be done
+				duration := time.Since(start).Seconds()
+				//Limiter started with 1 token, expecting a margin of error of 1
+				maxExpectedKept := 1 + int64(math.Ceil(duration)*100)
+
+				require.LessOrEqualf(t, kept, maxExpectedKept,
+					"Expected at most %d kept tokens for a %fs duration", maxExpectedKept, duration)
+			})
+		}
+
+		//Tests the limiter's ability to sample the traces when subjected sporadic bursts of requests.
+		//The limiter starts with a bucket filled with 100 tokens to handle a potential immediate first burst
+		for burstAmount := 10; burstAmount < 100; burstAmount += 10 {
+			t.Run(fmt.Sprintf("requests-bursts-%d-iterations", burstAmount), func(t *testing.T) {
+				burstFreq := 100 * time.Millisecond
+				burstSize := 10
```

Comment:
WDYT of testing with > 100 so that you can also assert on `skipped` too.

Reply:
We could set `burstSize` to 200 and check that  `skipped` and `kept` are both 100 at the end.